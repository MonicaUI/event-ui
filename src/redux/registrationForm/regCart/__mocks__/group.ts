/* eslint-env jest */
import noopAction from '../../../../testUtils/noopAction';

export const removeEventRegistrationFromRegCart = jest.fn(noopAction);

export const removeGroupMembersFromRegCart = jest.fn(noopAction);

export const addGroupMemberInRegCart = jest.fn(noopAction);

export const navigateToGroupMemberRegistration = jest.fn(noopAction);
