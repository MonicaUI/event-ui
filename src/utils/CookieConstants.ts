export const ALLOW = 'allow';
export const DENY = 'deny';
export const INITIALIZATION_BANNER = 'Initialization_Banner';
export const ALLPAGES_BANNER = 'AllPages_Banner';
export const REGISTRATION_CONVERSION_BANNER = 'RegistrationConversion_Banner';
export const OPT_IN_TYPE = 'opt-in';
export const BASIC_TYPE = 'basic';
export const INITIALIZATION = 'Initialization';
export const ALLPAGES = 'AllPages';
export const REGISTRATION_CONVERSION = 'RegistrationConversion';
export const TRIGGER_BASED = 'TriggerBased';
