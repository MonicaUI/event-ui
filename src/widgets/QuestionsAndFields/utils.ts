import { getIn } from 'icepick';
import getJSONValue from 'nucleus-widgets/utils/fields/getJSONValue';
import { getContactFieldForWidget } from 'event-widgets/redux/selectors/website/pluginData/registrationProcessNavigation';
import { getSubQuestionParentQuestionIdJsonPath } from 'event-widgets/redux/selectors/appData';
import { NAChoice, OtherChoice } from '@cvent/nucleus-form-validations';

const DEFAULT_QUESTION_CHOICES = Object.freeze([]);

export function getConfigWithDisplay(config: $TSFixMe, display: $TSFixMe): $TSFixMe {
  return {
    ...config,
    appData: {
      ...config.appData,
      display
    }
  };
}

export function convertChoicesToBaseLanguage(
  choices: $TSFixMe,
  choicesInCurrentLanguage: $TSFixMe,
  choicesInBaseLanguage: $TSFixMe
): $TSFixMe {
  return choices
    .map(choice => {
      const localizedChoice = choicesInCurrentLanguage.find(c => c.text === choice);
      if (localizedChoice) {
        const baseLangChoice = choicesInBaseLanguage.find(c => c.id === localizedChoice.id);
        if (baseLangChoice) {
          return {
            choice: baseLangChoice.text,
            answerType: 'Choice'
          };
        }
      } else {
        return {
          choice,
          answerType: 'Choice'
        };
      }
    })
    .filter(Boolean);
}

const emptyOrNaOrOther = value => !value || value === NAChoice || value === OtherChoice;

const needsTranslation = (baseChoices, localizedChoices, text) =>
  !localizedChoices.find(c => c.text === text) && baseChoices.find(c => c.text === text);

export function setSelectedValuesForChoices(
  answer: $TSFixMe,
  baseChoices: $TSFixMe,
  localizedChoices: $TSFixMe
): $TSFixMe {
  const translateEachIfNecessary = selectedValues =>
    selectedValues
      .filter(value => !emptyOrNaOrOther(value))
      .map(text =>
        needsTranslation(baseChoices, localizedChoices, text)
          ? updateSelectedValues(text, baseChoices, localizedChoices)
          : text
      );
  const selectedValues = translateEachIfNecessary(answer.value.selectedValues || []);
  const selectedPrevValues = translateEachIfNecessary(getIn(answer, ['valueBeforeMod', 'selectedValues']) || []);

  addMissingChoicesToLocaledAnswer(selectedValues, selectedPrevValues, answer);

  return {
    ...answer,
    value: {
      ...answer.value,
      selectedValues: selectedValues.length ? selectedValues : answer.value.selectedValues || []
    },
    valueBeforeMod: {
      ...answer.valueBeforeMod,
      selectedValues: selectedPrevValues.length
        ? selectedPrevValues
        : getIn(answer, ['valueBeforeMod', 'selectedValues']) || []
    }
  };
}

function addOtherOrNAChoiceToLocalizedAnswer(selectedValues, answerSelectedValues, otherOrNAChoice) {
  if (selectedValues.length && answerSelectedValues.includes(otherOrNAChoice)) {
    selectedValues.push(otherOrNAChoice);
  }
}

function addMissingChoicesToLocaledAnswer(selectedValues, selectedPrevValues, answer) {
  const answerSelectedValues = getIn(answer, ['value', 'selectedValues']) || [];
  const answerSelectedPrevValues = getIn(answer, ['valueBeforeMod', 'selectedValues']) || [];
  addOtherOrNAChoiceToLocalizedAnswer(selectedValues, answerSelectedValues, NAChoice);
  addOtherOrNAChoiceToLocalizedAnswer(selectedPrevValues, answerSelectedPrevValues, NAChoice);
  addOtherOrNAChoiceToLocalizedAnswer(selectedValues, answerSelectedValues, OtherChoice);
  addOtherOrNAChoiceToLocalizedAnswer(selectedPrevValues, answerSelectedPrevValues, OtherChoice);
}

function updateSelectedValues(choiceText, baseChoices, localizedChoices) {
  const choice = baseChoices.find(c => c.text === choiceText);
  if (choice) {
    const updatedChoice = localizedChoices.find(c => c.id === choice.id);
    if (updatedChoice) {
      return updatedChoice.text;
    }
  }
}

export function getLocalizedContactFieldForWidget(state: $TSFixMe, config: $TSFixMe, id: $TSFixMe): $TSFixMe {
  const registrationField = getContactFieldForWidget(state, config.registrationFieldPageType, config.fieldId, id);
  return {
    ...registrationField,
    displayName: getIn(config, ['appData', 'displayName']) || registrationField.displayName
  };
}

export function modifyFieldDefinitionForMultiLanguage(fieldDefinition: $TSFixMe): $TSFixMe {
  const modifiedFieldDefinition = {
    ...fieldDefinition,
    question: {
      ...fieldDefinition.question,
      questionTypeInfo: {
        ...fieldDefinition.question.questionTypeInfo,
        choices: fieldDefinition.question.questionTypeInfo.choices.map(choice => {
          return { id: choice.text.slice(choice.text.indexOf('.') + 1), text: choice.text };
        })
      }
    }
  };
  return modifiedFieldDefinition;
}

export function setParentQuestionId(appData: $TSFixMe, config: $TSFixMe, questionId: $TSFixMe): $TSFixMe {
  return {
    ...config,
    appData: {
      ...config.appData,
      parentQuestionId: getJSONValue(appData, getSubQuestionParentQuestionIdJsonPath(appData, questionId))
    }
  };
}

export function getQuestionCategory(question: $TSFixMe): $TSFixMe {
  if (question.isProductQuestion) {
    return 'productQuestions';
  }
  if (question.isTravelQuestion) {
    return 'travelQuestions';
  }
  return 'registrationQuestions';
}

export function setDefaultChoicesPreSelected(
  appData: $TSFixMe,
  questionCategory: $TSFixMe,
  questionId: $TSFixMe,
  answer: $TSFixMe,
  baseChoices: $TSFixMe
): $TSFixMe {
  const questionDefaultChoices = getIn(
    appData,
    ['registrationSettings', questionCategory, questionId, 'question', 'questionTypeInfo', 'defaultChoice'],
    DEFAULT_QUESTION_CHOICES
  );
  let filteredChoices = DEFAULT_QUESTION_CHOICES;
  if (questionDefaultChoices && questionDefaultChoices.length > 0) {
    if (!Array.isArray(questionDefaultChoices)) {
      filteredChoices = baseChoices.filter(c => c.id === questionDefaultChoices).map(choice => choice.text);
    } else {
      filteredChoices = baseChoices.map(choice => {
        const matchIndex = questionDefaultChoices.findIndex(defaultChoice => defaultChoice.id === choice.id);
        if (matchIndex !== -1) {
          return choice.text;
        }
      });
    }
    filteredChoices = filteredChoices.filter(x => x !== undefined);
    return {
      ...answer,
      value: {
        ...answer.value,
        selectedValues: filteredChoices
      },
      valueBeforeMod: {
        ...answer.valueBeforeMod,
        selectedValues: filteredChoices
      }
    };
  }
  return {
    ...answer,
    value: {
      ...answer.value,
      useDefaultChoices: false // answer = undefined and Question does not have defautl choices
    }
  };
}
