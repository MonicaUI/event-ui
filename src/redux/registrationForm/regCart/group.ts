import { setIn, getIn } from 'icepick';
import uuid from 'uuid';
import { hideLoadingOnError } from 'nucleus-guestside-site/src/redux/modules/dialogContainer';
import Logger from '@cvent/nucleus-logging';
import { loadAvailableCapacityCounts } from '../../capacity';
import {
  getPrimaryRegistrationId,
  getGuestsOfRegistrant,
  getRegistrationTypeId,
  getRegistrationPathId
} from './selectors';
import * as currentRegistrantSelectors from '../../selectors/currentRegistrant';
import {
  openCapacityReachedDialog,
  openGroupRegistrationTypeDialog,
  openKnownErrorDialog,
  openNoAdmissionItemAvailableForRegistrationTypeDialog
} from '../../../dialogs';
import { getEmailOnlyRegistrationTypes } from '../../selectors/shared';
import { getUpdateErrors, getModErrors, findKnownErrorResourceKey } from '../errors';
import { routeToFirstPageOfRegistrationForGroup } from '../../pathInfo';
import { getInitialProductRegistrations } from './internal';
import { startModification } from './workflow';
import {
  REMOVE_EVENT_REGISTRATION_ID,
  SET_CURRENT_EVENT_REGISTRATION_ID,
  ADD_GROUP_MEMBER_PENDING,
  ADD_GROUP_MEMBER_FAILURE,
  ADD_GROUP_MEMBER_SUCCESS
} from './actionTypes';
import { closeDialogContainer } from 'nucleus-guestside-site/src/redux/modules/dialogContainer';
import { updateGuestsInRegCart, removeGuestByEventRegistrationId } from './guests';
import { hasRegTypeCapacityWarning } from '../warnings';
const LOG = new Logger('redux/registrationForm/regCart/group');
import { getAvailableGroupRegTypeCapacities } from '../../../utils/regTypeCapacities';
import { defaultRegistrationTypeId } from 'event-widgets/utils/registrationType';
import {
  filterEventSnapshot,
  loadGuestRegistrationContent,
  loadRegistrationContent,
  evaluateQuestionVisibilityLogic
} from '../../actions';
import { isAdmissionItemsEnabled } from '../../selectors/event';
import { populateVisibleProducts } from '../../visibleProducts';
import { PENDING_APPROVAL, POST_REGISTRATION, REGISTRATION } from '../../website/registrationProcesses';
import { getRegistrationPathIdOrNull } from '../../selectors/currentRegistrationPath';
import { groupLeaderAndGroupMemberHaveCompatibleManualApprovalSettings } from '../../../utils/registrationUtils';
import { openPrivateEventErrorDialog } from '../../../dialogs';

function setGroupLeaderAttendeeType(regCart) {
  const groupLeaderEventRegId = getPrimaryRegistrationId(regCart);
  const regCartWithGroupLeader = setIn(
    regCart,
    ['eventRegistrations', groupLeaderEventRegId, 'attendeeType'],
    'GROUP_LEADER'
  );
  return regCartWithGroupLeader;
}

function updateRegCartWithNewEventRegistration(
  eventRegId,
  regCartClient,
  accessToken,
  state,
  inviteeId,
  contactId,
  eventId,
  isTestMode,
  referenceId,
  regTypeId,
  regCart
) {
  const regCartWithGroupLeader = setGroupLeaderAttendeeType(regCart);
  const regCartWithNewEventReg = setIn(regCartWithGroupLeader, ['eventRegistrations', eventRegId], {
    eventId,
    eventRegistrationId: eventRegId,
    attendee: {
      personalInformation: {},
      eventAnswers: {}
    },
    attendeeType: 'ATTENDEE',
    primaryRegistrationId: getPrimaryRegistrationId(regCart),
    registrationTypeId: regTypeId,
    productRegistrations: getInitialProductRegistrations(state, regTypeId)
  });
  LOG.debug('added new event reg for group member', regCart);
  return regCartWithNewEventReg;
}

export function removeEventRegistrationFromRegCart(currentEventRegistrationId: $TSFixMe): $TSFixMe {
  return {
    type: REMOVE_EVENT_REGISTRATION_ID,
    payload: {
      currentEventRegistrationId,
      path: ['eventRegistrations']
    }
  };
}

export function removeGroupMembersFromRegCart(eventRegIds: $TSFixMe) {
  return async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      registrationForm: { regCart }
    } = getState();
    if (currentRegistrantSelectors.isGroupRegistration(getState()) && regCart && regCart.eventRegistrations) {
      let updatedRegCart = regCart;
      for (const eventRegId of eventRegIds) {
        const guests = getGuestsOfRegistrant(regCart, eventRegId);
        /* If the GM has guest remove them then update the regCart */
        if (guests.length !== 0) {
          for (const eventReg of guests) {
            updatedRegCart = await removeGuestByEventRegistrationId(updatedRegCart, eventReg.eventRegistrationId);
          }
        }
        // remove the attendee
        updatedRegCart = await removeGuestByEventRegistrationId(updatedRegCart, eventRegId);
      }
      await dispatch(updateGuestsInRegCart(updatedRegCart));
    }
  };
}

const applyGroupMemberRegistrationTypeSelection = (
  regCartWithGroupMember,
  groupMemberEventRegId,
  inviteeId,
  contactId
) => {
  return async (dispatch, getState) => {
    dispatch(closeDialogContainer());
    const chosenRegistrationType = getIn(regCartWithGroupMember, [
      'eventRegistrations',
      groupMemberEventRegId,
      'registrationTypeId'
    ]);
    if (canAddGroupMemberFromCurrentPage(getState())) {
      await dispatch(startModification());

      const manualApprovalConfig = { knownMemberRegTypeId: chosenRegistrationType };
      if (!groupLeaderAndGroupMemberHaveCompatibleManualApprovalSettings(getState(), manualApprovalConfig)) {
        return dispatch(openPrivateEventErrorDialog());
      }

      const {
        accessToken,
        clients: { regCartClient },
        userSession: { referenceId },
        defaultUserSession: { isTestMode },
        registrationForm: { regCart },
        event
      } = getState();
      const eventId = event.id;

      let regModCartWithGroupMember = updateRegCartWithNewEventRegistration(
        groupMemberEventRegId,
        regCartClient,
        accessToken,
        getState(),
        inviteeId,
        contactId,
        eventId,
        isTestMode,
        referenceId,
        chosenRegistrationType,
        regCart
      );

      regModCartWithGroupMember = setIn(
        regModCartWithGroupMember,
        ['eventRegistrations', groupMemberEventRegId, 'registrationTypeId'],
        chosenRegistrationType
      );

      await dispatch(continueWithAddGroupMember(regModCartWithGroupMember, groupMemberEventRegId));
    } else {
      const updatedRegCart = setIn(
        regCartWithGroupMember,
        ['eventRegistrations', groupMemberEventRegId, 'productRegistrations'],
        getInitialProductRegistrations(getState(), chosenRegistrationType)
      );
      await dispatch(continueWithAddGroupMember(updatedRegCart, groupMemberEventRegId));
    }
  };
};

export function getVisibleRegTypes(
  openRegTypes: $TSFixMe,
  state: $TSFixMe,
  isAdmissionItemsEnabledFlag: $TSFixMe,
  regTypeHasAvailableAdmissionItemMap: $TSFixMe
): $TSFixMe {
  const emailOnlyTypes = getEmailOnlyRegistrationTypes(state);
  const currentRegistrationTypeId = currentRegistrantSelectors.getRegistrationTypeId(state);
  return openRegTypes.filter(regType => {
    const isEmailOnly = emailOnlyTypes.includes(regType.id);
    const hasAvailableAdmission = isAdmissionItemsEnabledFlag ? regTypeHasAvailableAdmissionItemMap[regType.id] : true;
    // If the current registrant's type is a private type and has admission available, show it as an option.
    if (regType.name !== '' && isEmailOnly && hasAvailableAdmission && regType.id === currentRegistrationTypeId)
      return true;
    // If the reg type is not associated with an private reg path and has admission available, show it as an option.
    if (regType.name !== '' && !isEmailOnly && hasAvailableAdmission) return true;
  });
}

function canAddGroupMemberFromCurrentPage(state) {
  return POST_REGISTRATION.isTypeOfCurrentPage(state) || PENDING_APPROVAL.isTypeOfCurrentPage(state);
}

export function addGroupMemberInRegCart(eventId: $TSFixMe, inviteeId?: $TSFixMe, contactId?: $TSFixMe) {
  return async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      appData: {
        registrationSettings: { showModalForGroupMemberRegistrationTypes }
      }
    } = getState();
    const capacityInfo = getAvailableGroupRegTypeCapacities(getState());
    const isRegTypesEnabled = getIn(getState(), [
      'event',
      'eventFeatureSetup',
      'registrationProcess',
      'multipleRegistrationTypes'
    ]);
    const regTypes = capacityInfo.regTypeCapacitiesAvailable.map(regType => ({
      ...getState().event.registrationTypes[regType.id],
      closedForGroupMember: isRegTypesEnabled && regType.available < 1
    }));
    const openRegTypes = regTypes.filter(regType => !regType.closedForGroupMember);
    // add check to make sure published in phase 5
    const regTypeHasAvailableAdmissionItemMap = currentRegistrantSelectors.getRegTypeHasAvailableAdmissionItemMap(
      getState()
    );
    const isAdmissionItemsEnabledFlag = isAdmissionItemsEnabled(getState());
    const visibleRegTypes = getVisibleRegTypes(
      openRegTypes,
      getState(),
      isAdmissionItemsEnabledFlag,
      regTypeHasAvailableAdmissionItemMap
    );
    const displayRegTypeModal = visibleRegTypes.length > 1 && showModalForGroupMemberRegistrationTypes;
    if (canAddGroupMemberFromCurrentPage(getState()) && !displayRegTypeModal) {
      try {
        await dispatch(startModification());
      } catch (error) {
        if (getModErrors.isKnownError(error)) {
          return await dispatch(openKnownErrorDialog(findKnownErrorResourceKey(error.responseBody.validationMessages)));
        } else if (!groupLeaderAndGroupMemberHaveCompatibleManualApprovalSettings(getState())) {
          return dispatch(openPrivateEventErrorDialog());
        }
      }
    }
    const {
      accessToken,
      clients: { regCartClient },
      userSession: { referenceId },
      defaultUserSession: { isTestMode },
      registrationForm: { regCart }
    } = getState();
    /**
     * If there is only one visible regType to the member, we automatically give them that,
     * if not, we give the default regType and dispatch the regtype selection modal.
     */
    let regTypeIdToSet = defaultRegistrationTypeId;
    if (openRegTypes && openRegTypes.length > 0 && visibleRegTypes.length < 2) {
      regTypeIdToSet = (visibleRegTypes.length === 0 ? openRegTypes : visibleRegTypes)[0].id;
    }
    const eventRegId = String(uuid.v4());
    const regCartWithGroupMember = await updateRegCartWithNewEventRegistration(
      eventRegId,
      regCartClient,
      accessToken,
      getState(),
      inviteeId,
      contactId,
      eventId,
      isTestMode,
      referenceId,
      regTypeIdToSet,
      regCart
    );

    if (displayRegTypeModal) {
      return dispatch(
        openGroupRegistrationTypeDialog(
          visibleRegTypes,
          regCartWithGroupMember,
          eventRegId,
          applyGroupMemberRegistrationTypeSelection,
          inviteeId,
          contactId,
          regTypeHasAvailableAdmissionItemMap,
          isAdmissionItemsEnabledFlag
        )
      );
    }
    await dispatch(continueWithAddGroupMember(regCartWithGroupMember, eventRegId));
  };
}

const setCurrentEventRegistrationId = eventRegistrationId => dispatch => {
  dispatch({
    type: SET_CURRENT_EVENT_REGISTRATION_ID,
    payload: {
      currentEventRegistrationId: eventRegistrationId
    }
  });
};

function continueWithAddGroupMember(regCartWithGroupMember, groupMemberEventRegId) {
  return async (dispatch, getState) => {
    dispatch({
      type: ADD_GROUP_MEMBER_PENDING
    });
    let response;
    const {
      accessToken,
      clients: { regCartClient },
      text: { translate },
      event,
      account
    } = getState();
    try {
      response = await regCartClient.updateRegCart(accessToken, regCartWithGroupMember);
      dispatch({
        type: ADD_GROUP_MEMBER_SUCCESS,
        payload: {
          regCart: response.regCart,
          validationMessages: response.validationMessages,
          currentEventRegistrationId: groupMemberEventRegId
        }
      });
      const regCart = response.regCart.eventRegistrations ? response.regCart : regCartWithGroupMember;
      // filter snaphsot to show updated prices for products for the group member
      const { registrationTypeId, registrationPathId } = regCart.eventRegistrations[groupMemberEventRegId];
      await dispatch(filterEventSnapshot(getState().eventSnapshotVersion, registrationTypeId, registrationPathId));
      await dispatch(loadAvailableCapacityCounts());
      LOG.debug('addGroupMember success');
    } catch (error) {
      LOG.info('addGroupMember failed', error);
      // if we get external auth or oauth error , we need to redirect to external auth or oauth url
      if (
        getUpdateErrors.handleAuthError(
          error,
          account,
          event,
          getRegistrationTypeId(regCartWithGroupMember, groupMemberEventRegId),
          getRegistrationPathId(regCartWithGroupMember, groupMemberEventRegId)
        )
      ) {
        return;
      }
      dispatch({ type: ADD_GROUP_MEMBER_FAILURE, payload: { error } });
      dispatch(hideLoadingOnError());
      if (getUpdateErrors.isAddGroupMemberNotAvailableError(error)) {
        await dispatch(loadAvailableCapacityCounts());
        return await dispatch(
          openCapacityReachedDialog({
            subMessage: translate('EventGuestSide_GroupMember_CapacityReachedError__resx')
          })
        );
      }
      if (getUpdateErrors.isRegTypeInvalidForGroup(error)) {
        const subMessage = 'EventGuestSide_RegistrationTypeError_NoRegistrationTypeAvailableHelpText_resx';
        const message = 'EventGuestSide_RegistrationTypeConflict_Title_resx';
        return await dispatch(openKnownErrorDialog(subMessage, message));
      }
      if (getUpdateErrors.isAdmissionItemsNotAvailableForRegTypeError(error)) {
        const callBack = () => {
          const state = getState();
          const capacityInfo = getAvailableGroupRegTypeCapacities(state);
          const regTypes = capacityInfo.regTypeCapacitiesAvailable.map(regType => ({
            ...state.event.registrationTypes[regType.id],
            closedForGroupMember: regType.available < 1
          }));
          return dispatch(
            openGroupRegistrationTypeDialog(
              regTypes,
              regCartWithGroupMember,
              groupMemberEventRegId,
              applyGroupMemberRegistrationTypeSelection
            )
          );
        };
        return dispatch(openNoAdmissionItemAvailableForRegistrationTypeDialog(callBack));
      }
      throw error;
    }
    if (hasRegTypeCapacityWarning(getState())) {
      await dispatch(loadAvailableCapacityCounts());
      dispatch(setCurrentEventRegistrationId(getPrimaryRegistrationId(regCartWithGroupMember)));
      await dispatch(removeGroupMembersFromRegCart([groupMemberEventRegId]));
      return await dispatch(
        openCapacityReachedDialog({
          subMessage: getState().text.translate('EventGuestSide_RegType_CapacityReachedSubMessage__resx')
        })
      );
    }
    await Promise.all([
      dispatch(populateVisibleProducts()),
      dispatch(loadRegistrationContent(REGISTRATION, getRegistrationPathIdOrNull(getState()))),
      dispatch(loadGuestRegistrationContent(getRegistrationPathIdOrNull(getState()))),
      dispatch(evaluateQuestionVisibilityLogic(null, true))
    ]);
    await dispatch(
      routeToFirstPageOfRegistrationForGroup(
        groupMemberEventRegId,
        REGISTRATION.forCurrentRegistrant().startPageId(getState())
      )
    );
    return response;
  };
}

export const navigateToGroupMemberRegistration =
  (eventRegistrationId: $TSFixMe): $TSFixMe =>
  async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    dispatch(setCurrentEventRegistrationId(eventRegistrationId));
    const {
      registrationForm: { regCart }
    } = getState();
    // filter snaphsot to show updated prices for products for the group member
    const { registrationTypeId, registrationPathId } = regCart.eventRegistrations[eventRegistrationId];
    await Promise.all([
      dispatch(filterEventSnapshot(getState().eventSnapshotVersion, registrationTypeId, registrationPathId)),
      dispatch(loadRegistrationContent(REGISTRATION, getRegistrationPathIdOrNull(getState())))
    ]);
    await dispatch(
      routeToFirstPageOfRegistrationForGroup(
        eventRegistrationId,
        REGISTRATION.forCurrentRegistrant().startPageId(getState())
      )
    );
  };
