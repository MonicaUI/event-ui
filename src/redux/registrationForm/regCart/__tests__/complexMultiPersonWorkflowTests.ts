import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import EventSnapshot from '../../../../../fixtures/EventSnapshot.json';
import transformEventData from 'event-widgets/utils/transformEventData';
import { setIn } from 'icepick';
import { addGroupMemberInRegCart } from '../group';
import {
  createAdmissionItem,
  createRegistrationType,
  createAutoAppliedAdmissionItem
} from '../../../../testUtils/index';
import { defaultRegistrationTypeId } from 'event-widgets/utils/registrationType';

const PRIMARY_INVITEE_REGISTRATION_ID = 'primaryEventRegId';
const GROUP_MEMBER_REGISTRATION_ID = 'groupMemberEventRegId';
const accessToken = '';

jest.mock('uuid', () => {
  return {
    v4: jest.fn().mockReturnValue('groupMemberEventRegId')
  };
});

jest.mock('../../../actions', () => {
  return {
    ...jest.requireActual<$TSFixMe>('../../../actions'),
    filterEventSnapshot: () => () => {},
    loadRegistrationContent: () => () => {},
    loadGuestRegistrationContent: () => () => {}
  };
});

const regCartClient = {
  updateRegCart: jest.fn(() => {
    return {
      regCart: {}
    };
  })
};
const capacityClient = {
  getCapacitySummaries: jest.fn(() => {})
};
const eventEmailClient = {};

const eventSnapshotClient = {
  getAccountSnapshot: jest.fn(() => EventSnapshot.accountSnapshot),
  getEventSnapshot: jest.fn(() => EventSnapshot.eventSnapshot)
};
const productVisibilityClient = {
  getVisibleProducts: jest.fn(() => ({ Sessions: {} }))
};
const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

const defaultRegCart = {
  regCartId: 'regCartAId',
  eventRegistrations: {
    [PRIMARY_INVITEE_REGISTRATION_ID]: {
      eventRegistrationId: PRIMARY_INVITEE_REGISTRATION_ID,
      eventId: 'EventAId',
      registrationTypeId: defaultRegistrationTypeId,
      attendee: {
        attendeeId: 'attendeeAId',
        personalInformation: {
          firstName: 'floofykins',
          lastName: 'snuffles',
          emailAddress: 'floofykins.snuffles@j.mail',
          socialMediaUrls: {
            FACEBOOK: 'http://www.facebook.com/YourChosenName'
          },
          customFields: {
            customFieldAId: {
              questionId: 'customFieldAId',
              answers: [
                {
                  answerType: 'I am a dragon!',
                  text: 'text answer'
                }
              ]
            }
          }
        },
        eventAnswers: {
          eventAnswerAId: {
            questionId: 'eventAnswerAId',
            answers: [
              {
                answerType: 'Im Hungry!',
                text: 'text answer'
              }
            ]
          }
        }
      },
      attendeeType: 'ATTENDEE',
      productRegistrations: [
        {
          productId: 'admissionItemAId',
          productType: 'AdmissionItem',
          quantity: 1,
          requestedAction: 'REGISTER'
        }
      ],
      sessionRegistrations: {
        sessionAId: {
          productId: 'sessionAId',
          productType: 'Session',
          quantity: 1,
          requestedAction: 'REGISTER'
        }
      }
    }
  }
};

const initialState = {
  accessToken,
  website: EventSnapshot.eventSnapshot.siteEditor.website,
  appData: transformEventData(
    EventSnapshot.eventSnapshot.siteEditor.eventData,
    EventSnapshot.accountSnapshot,
    EventSnapshot.eventSnapshot,
    EventSnapshot.eventSnapshot.siteEditor.website
  ),
  account: {
    settings: {
      dupMatchKeyType: 'EMAIL_ONLY'
    }
  },
  event: {
    eventFeatureSetup: {
      fees: {
        fees: true
      }
    },
    registrationTypes: EventSnapshot.eventSnapshot.registrationTypes,
    products: EventSnapshot.eventSnapshot.products,
    id: EventSnapshot.eventSnapshot.id,
    capacityId: 'event_capacity'
  },
  registrantLogin: {
    form: {
      firstName: 'firstName',
      lastName: 'lastName',
      emailAddress: 'emailAddress',
      confirmationNumber: 'confirmationNumber'
    },
    status: {
      login: {},
      resendConfirmation: {}
    }
  },
  text: {
    translate: (resx, tokens) => (tokens ? `${resx}:${JSON.stringify(tokens)}` : resx)
  },
  clients: { regCartClient, capacityClient, eventEmailClient, eventSnapshotClient, productVisibilityClient },
  userSession: {},
  defaultUserSession: {
    isPreview: false
  },
  registrationForm: {
    currentEventRegistrationId: PRIMARY_INVITEE_REGISTRATION_ID,
    regCart: defaultRegCart
  },
  pathInfo: {
    rootPath: 'events/dummyId',
    currentPageId: 'regProcessStep1'
  }
};

let store;
let regCartUtils;
let currentRegistrationPathSelectors;
beforeEach(() => {
  jest.clearAllMocks();
  store = mockStore(initialState);
  regCartUtils = require('../../../../utils/regTypeCapacities');
  currentRegistrationPathSelectors = require('../../../selectors/currentRegistrationPath');
});

describe('limit group member reg type tests', () => {
  test('correct regTypeId and admissionItem are set if only one regType is availble to group members', async () => {
    const regTypeA = createRegistrationType('regTypeA');
    const regTypeB = createRegistrationType('regTypeB');
    const admItemA = createAdmissionItem('AdmissionItemAId', [defaultRegistrationTypeId, regTypeA.id]);
    const defaultAdmissionItem = Object.values(initialState.event.products.admissionItems)[0];

    const updatedEventSnapshot = {
      ...initialState.event,
      registrationTypes: {
        ...initialState.event.registrationTypes,
        [regTypeA.id]: regTypeA,
        [regTypeB.id]: regTypeB
      },
      products: {
        admissionItems: {
          [defaultAdmissionItem.id]: {
            ...defaultAdmissionItem,
            applicableContactTypes: [regTypeB.id]
          },
          [admItemA.id]: admItemA
        }
      }
    };

    regCartUtils.getAvailableGroupRegTypeCapacities = jest.fn(() => {
      return {
        regTypeCapacitiesAvailable: [{ id: regTypeB.id, available: Infinity }]
      };
    });

    currentRegistrationPathSelectors.getRegistrationStartPageId = jest.fn(() => 'regProcessStep1');
    const updatedInitialState = setIn(initialState, ['event'], updatedEventSnapshot);
    const regCart = initialState.registrationForm.regCart;
    /**
     * Group members can only get regTypeB and the only visible admission item to regTypeB is the default one
     *
     */
    const expectedRegCart = {
      ...regCart,
      eventRegistrations: {
        [PRIMARY_INVITEE_REGISTRATION_ID]: {
          ...regCart.eventRegistrations[PRIMARY_INVITEE_REGISTRATION_ID],
          attendeeType: 'GROUP_LEADER'
        },
        [GROUP_MEMBER_REGISTRATION_ID]: {
          eventId: initialState.event.id,
          eventRegistrationId: GROUP_MEMBER_REGISTRATION_ID,
          attendee: {
            personalInformation: {},
            eventAnswers: {}
          },
          attendeeType: 'ATTENDEE',
          primaryRegistrationId: PRIMARY_INVITEE_REGISTRATION_ID,
          productRegistrations: [createAutoAppliedAdmissionItem(defaultAdmissionItem)],
          registrationTypeId: regTypeB.id
        }
      }
    };
    store = mockStore(updatedInitialState);
    const eventId = initialState.event.id;
    await store.dispatch(addGroupMemberInRegCart(eventId));
    expect(regCartClient.updateRegCart).toHaveBeenCalledWith(accessToken, expectedRegCart);
    expect(store.getActions()).toMatchSnapshot();
  });
});
