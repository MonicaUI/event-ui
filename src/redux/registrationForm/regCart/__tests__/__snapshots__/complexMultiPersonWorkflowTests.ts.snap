// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`limit group member reg type tests correct regTypeId and admissionItem are set if only one regType is availble to group members 1`] = `
Array [
  Object {
    "type": "event-guestside-site/regCart/ADD_GROUP_MEMBER_PENDING",
  },
  Object {
    "payload": Object {
      "currentEventRegistrationId": "groupMemberEventRegId",
      "regCart": Object {},
      "validationMessages": undefined,
    },
    "type": "event-guestside-site/regCart/ADD_GROUP_MEMBER_SUCCESS",
  },
  Object {
    "payload": Object {
      "capacity": undefined,
    },
    "type": "event-widgets/capacity/LOAD_CAPACITY",
  },
  Object {
    "payload": Object {
      "questionIds": Array [],
      "visibility": Object {},
    },
    "type": "event-guestside-site/UPDATE_QUESTION_VISIBILITY",
  },
  Object {
    "payload": Object {
      "eventRegistrationId": "primaryEventRegId",
      "visibleProducts": Object {
        "Sessions": Object {},
      },
    },
    "type": "event-guestside-site/LOAD_VISIBLE_SESSION_PRODUCTS",
  },
]
`;
